packages/webapp/src/containers/Animals/Inventory/useAnimalOrBatchRemoval.ts

```diff
@@ -173,31 +177,52 @@ const useAnimalOrBatchRemoval = (
     return new Set([...completedTasks, ...abandonedTasks].map(({ task_id }) => task_id));
   }, [completedTasks, abandonedTasks]);
 
-  const hasFinalizedTasks = useMemo(() => {
-    if (!removalModalOpen) {
-      return false;
-    }
-
-    const selectedInventoryIdsSet = new Set(selectedInventoryIds);
-    const finalizedTaskIdsSet = getFinalizedTaskIdsSet();
+  const checkHasFinalizedTasks = async () => {
+    const animalIds: number[] = [];
+    const selectedAnimalIds: string[] = [];
+    const animalBatchIds: number[] = [];
+    const selectedBatchIds: string[] = [];
+    let result;
 
-    let inventoryFoundCount = 0;
-    for (let { id, tasks } of animalTasksWithInventoryIds) {
-      if (selectedInventoryIdsSet.has(id)) {
-        inventoryFoundCount++;
+    for (const id of selectedInventoryIds) {
+      const { kind, id: entity_id } = parseInventoryId(id);
+      if (kind === AnimalOrBatchKeys.ANIMAL) {
+        animalIds.push(entity_id);
+        selectedAnimalIds.push(id);
+      } else if (AnimalOrBatchKeys.BATCH) {
+        animalBatchIds.push(entity_id);
+        selectedBatchIds.push(id);
+      }
+    }
 
-        if (tasks.some(({ task_id }) => finalizedTaskIdsSet.has(task_id))) {
-          return true;
-        }
+    if (animalIds.length) {
+      result = await mutations['checkDeleteAnimals'].trigger(animalIds);
+      if (result.error.data === 'Cannot delete animal with completed or abandoned tasks') {
+        return true;
       }
-      // Stop iterating once all selected inventory IDs have been processed.
-      if (inventoryFoundCount === selectedInventoryIdsSet.size) {
-        return false;
+    }
+    if (animalBatchIds.length) {
+      result = await mutations['checkDeleteAnimalBatches'].trigger(animalBatchIds);
+      if (result.error.data === 'Cannot delete animal with completed or abandoned tasks') {
+        return true;
       }
     }
-  }, [removalModalOpen, completedTasks, abandonedTasks, selectedInventoryIds]);
+    return false;
+  };
 
-  return { onConfirmRemoveAnimals, removalModalOpen, setRemovalModalOpen, hasFinalizedTasks };
+  const onClickRemoveAnimals = async () => {
+    const hideDelete = await checkHasFinalizedTasks();
+    setHasFinalizedTasks(hideDelete);
+    setRemovalModalOpen(true);
+  };
```

Comment:
I think the ideal behaviour is to open the modal first and handle the loading status within the modal. The APIs seem fast enough, though.

Reply:
I'm not sure ! I figured I would give it the extra millisecond to run the query haha

But that makes sense too.. should I update?